rbs/MethodTypeToParserNode.cc

```diff
@@ -69,6 +81,69 @@ core::NameRef nodeName(const parser::Node *node) {
     return name;
 }
 
+string argKindToString(RBSArg::Kind kind) {
+    switch (kind) {
+        case RBSArg::Kind::Positional:
+            return "positional";
+        case RBSArg::Kind::OptionalPositional:
+            return "optional positional";
+        case RBSArg::Kind::RestPositional:
+            return "rest positional";
+        case RBSArg::Kind::Keyword:
+            return "keyword";
+        case RBSArg::Kind::OptionalKeyword:
+            return "optional keyword";
+        case RBSArg::Kind::RestKeyword:
+            return "rest keyword";
+        case RBSArg::Kind::Block:
+            return "block";
+    }
+}
+
+string nodeKindToString(const parser::Node *node) {
+    string kind;
+
+    typecase(
+        node, [&](const parser::Arg *parserArg) { kind = "positional"; },
+        [&](const parser::Optarg *parserArg) { kind = "optional positional"; },
+        [&](const parser::Restarg *parserArg) { kind = "rest positional"; },
+        [&](const parser::Kwarg *parserArg) { kind = "keyword"; },
+        [&](const parser::Kwoptarg *parserArg) { kind = "optional keyword"; },
+        [&](const parser::Kwrestarg *parserArg) { kind = "rest keyword"; },
+        [&](const parser::Blockarg *parserArg) { kind = "block"; },
+        [&](const parser::Node *other) {
+            Exception::raise("Unexpected expression type: {}", ((parser::Node *)node)->nodeName());
+        });
+
+    return kind;
+}
+
+bool checkParameterKind(core::MutableContext ctx, const RBSDeclaration &declaration, const RBSArg &arg,
+                        const parser::Node *methodArg) {
+    auto error = false;
+
+    typecase(
+        methodArg, [&](const parser::Arg *parserArg) { error = arg.kind != RBSArg::Kind::Positional; },
+        [&](const parser::Optarg *parserArg) { error = arg.kind != RBSArg::Kind::OptionalPositional; },
+        [&](const parser::Restarg *parserArg) { error = arg.kind != RBSArg::Kind::RestPositional; },
+        [&](const parser::Kwarg *parserArg) { error = arg.kind != RBSArg::Kind::Keyword; },
+        [&](const parser::Kwoptarg *parserArg) { error = arg.kind != RBSArg::Kind::OptionalKeyword; },
+        [&](const parser::Kwrestarg *parserArg) { error = arg.kind != RBSArg::Kind::RestKeyword; },
+        [&](const parser::Blockarg *parserArg) { error = arg.kind != RBSArg::Kind::Block; },
+        [&](const parser::Node *other) {
+            Exception::raise("Unexpected expression type: {}", ((parser::Node *)methodArg)->nodeName());
+        });
+
+    if (error) {
+        if (auto e = ctx.beginError(arg.loc, core::errors::Rewriter::RBSIncorrectParameterKind)) {
+            e.setHeader("Argument kind mismatch for `{}`, expected `{}`, got `{}`", nodeName(methodArg).show(ctx.state),
```

Comment:
The error message isn't super clear to me; it's not obvious that `expected` is the arg in the `def`, and `got` is the one in the `sig`. Maybe something like:

> error: Argument kind mismatch for `p`, method declares `positional`, but RBS sig declares `optional positional`